Tests/SwiftFusionTests/Inference/GenericFactorTests.swift

```diff
@@ -19,57 +19,87 @@ import XCTest
   /// Tests errors in a simple factor graph.
   func testErrors() {
-    // Set up 4 variables:
-    // - `intVar1` is an `Int` with value `0`.
-    // - `intVar2` is an `Int` with value `1`.
-    // - `poseVar1` is a `Pose2` with value `Pose2(1, 0, 0)`.
-    // - `poseVar2` is a `Pose2` with value `Pose2(0, 1, 1)`.
+    // A switching linear motion factor graph.
+    //
+    // There are 3 pose variables and 2 motion labels in the graph. The initial guesses are:
+    // - motion labels: 0, 1
+    // - poses: Pose2(0, 0, 0), Pose2(1, 1, 0), Pose2(1, 1, 0.9)
+    //
+    // The factors are:
+    // - a prior factor giving a prior of `Pose2(0, 0, 0)` on the first position.
+    // - two motion factors with possible motions [Pose2(1, 1, 0), Pose2(0, 0, 1)]
+
+    // Set up the initial guess.
 
     let intVariables = ArrayStorage<Int>.create(minimumCapacity: 2)
-    let intVar1 = TypedID<Int, Int>(intVariables.append(0)!)
-    let intVar2 = TypedID<Int, Int>(intVariables.append(1)!)
+    let motionLabel1ID = TypedID<Int, Int>(intVariables.append(0)!)
+    let motionLabel2ID = TypedID<Int, Int>(intVariables.append(1)!)
 
-    let poseVariables = ArrayStorage<Pose2>.create(minimumCapacity: 2)
-    let poseVar1 = TypedID<Pose2, Int>(poseVariables.append(Pose2(1, 0, 0))!)
-    let poseVar2 = TypedID<Pose2, Int>(poseVariables.append(Pose2(0, 1, 1))!)
+    let poseVariables = ArrayStorage<Pose2>.create(minimumCapacity: 3)
+    let pose1ID = TypedID<Pose2, Int>(poseVariables.append(Pose2(0, 0, 0))!)
+    let pose2ID = TypedID<Pose2, Int>(poseVariables.append(Pose2(1, 1, 0))!)
+    let pose3ID = TypedID<Pose2, Int>(poseVariables.append(Pose2(1, 1, 0.9))!)
 
     let variableAssignments = ValuesArray(contiguousStorage: [
       ObjectIdentifier(Int.self): intVariables,
       ObjectIdentifier(Pose2.self): poseVariables
     ])
 
-    // Set up 4 `DemoFactor`s, connected to each combination of `Int` and `Pose2` variables.
+    // Set up the factor graph.
 
-    let priors = [Pose2(1, 0, 0), Pose2(0, 1, 1)]
-    let factors = FactorArrayStorage<DemoFactor<Pose2>>.create(minimumCapacity: 4)
-    _ = factors.append(DemoFactor(edges: Tuple2(intVar1, poseVar1), priors: priors))
-    _ = factors.append(DemoFactor(edges: Tuple2(intVar1, poseVar2), priors: priors))
-    _ = factors.append(DemoFactor(edges: Tuple2(intVar2, poseVar1), priors: priors))
-    _ = factors.append(DemoFactor(edges: Tuple2(intVar2, poseVar2), priors: priors))
+    let priorFactors = FactorArrayStorage<PriorFactor<Pose2>>.create(minimumCapacity: 1)
+    _ = priorFactors.append(PriorFactor(edges: Tuple1(pose1ID), prior: Pose2(0, 0, 0)))
+
+    let motionFactors =
+      FactorArrayStorage<SwitchingLinearMotionFactor<Pose2>>.create(minimumCapacity: 2)
+    _ = motionFactors.append(SwitchingLinearMotionFactor(
+      edges: Tuple3(motionLabel1ID, pose1ID, pose2ID),
+      motions: [Pose2(1, 1, 0), Pose2(0, 0, 1)]
+    ))
+    _ = motionFactors.append(SwitchingLinearMotionFactor(
```

Comment:
no such thing as "emplace" I guess?

Reply:
No such thing, and I think Swift's value semantics make it so that "append" is just as fast as "emplace" would be. Though I'm not sure, because I'm not very knowledgeable about value semantics and "emplace". @dabrahams might know for sure.

Reply:
The feature of emplace that I like is that we don't have to repeat the constructor here, rather the constructor arguments are forwarded, yielding less clutter - the collection knows the type, yes?

Reply:
Yeah, that doesn't exist, at least not automatically.

There is a way to manually define `emplace` for collections that know their element type, but you have to write separate code for each element type:

```
extension RangeReplacableCollection {
  func emplace<Pose: LieGroup>(edges: Element.Edges, motions: [Pose]) where Element == SwitchingMotionModelFactor<Pose> {
    append(SwitchingMotionModelFactor(edges: edges, motions: motions)
  }
}
```

Probably not worth it if we have to do it for each factor? Anyways, I'm not going to do anything about this now because I'm focusing on getting the functionality working without any sugar.

Reply:
You can just replace `SwitchingLinearMotionFactor` with `.init`.